### PR TITLE
Add support for mixed precision in contraction scale and bilinear

### PR DESCRIPTION
**Pull request by bwroblew**
This PR introduces support for mixed precision datatypes in the contraction operation.
It exposes a new template argument - `ComputeDataType` that is used to cast the data before computations.

The combinations supported after the change:
| Input/Output datatype | Compute datatype |
| ------------------------- | --------------------- |
| fp32  | fp32 |
| fp64  | fp64 |
| fp64 | fp32 |
| fp32 | fp16 |
| fp32 | bf16 |
| fp16 | fp32 |
| bf16 | fp32 |

This change also fixes a lot of bugs/issues I found in the reference contraction, profiler and tests:
- incorrect use of `type_convert` with `const` datatypes in reference contraction;
- lack of cast to the dst datatype in reference contraction;
- incorrect order of dimensions of matrix B passed to the reference contraction;
- incorrect order of strides of matrix B passed to the contraction kernel;
- incorrectly calculated tolerance passed to `check_err` for BF16 and F16 datatypes;
- tests not taking into account the memory layouts set in test cases - testing only two memory layouts hardcoded in strides;
- profiler passing input datatype to reference contraction instead of the accumulation datatype.

Normally I would prepare a separate review for every issue fix, but since GitHub does not allow for dependent PRs, it would take ages to finally merge this change. That's why I decided to put it all in this single huuuge change (it's mostly huge since I'm adding a lot of new instances, fixes themselves are mostly 1-5 liners).

### CODE REVIEW

File: library/src/tensor_operation_instance/gpu/contraction_bilinear/device_contraction_bilinear_m2_n2_k2_xdl_c_shuffle_bf16_bf16_bf16_bf16_compute_f32_kknn_instance.cpp [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+#include "ck/tensor_operation/gpu/element/element_wise_operation.hpp"
+#include "ck/library/tensor_operation_instance/gpu/contraction/device_contraction_instance.hpp"
+#include "ck/library/tensor_operation_instance/add_device_operation_instance.hpp"
+
+namespace ck {
+namespace tensor_operation {
+namespace device {
+namespace instance {
+
+using BF16       = ck::bhalf_t;
```
Comment (bartekxk) [inline library/src/tensor_operation_instance/gpu/contraction_bilinear/device_contraction_bilinear_m2_n2_k2_xdl_c_shuffle_bf16_bf16_bf16_bf16_compute_f32_kknn_instance.cpp]: You can cut these lines to device_contraction_instance.hpp file from each cpp instance file.

**Review by bartekxk — DISMISSED** [review]
LGTM

File: library/src/tensor_operation_instance/gpu/contraction_bilinear/device_contraction_bilinear_m2_n2_k2_xdl_c_shuffle_bf16_bf16_bf16_bf16_compute_f32_kknn_instance.cpp (hunk 1 above)
Comment (bwroblew) [reply library/src/tensor_operation_instance/gpu/contraction_bilinear/device_contraction_bilinear_m2_n2_k2_xdl_c_shuffle_bf16_bf16_bf16_bf16_compute_f32_kknn_instance.cpp]: Moved them from here and from examples

File: profiler/include/profiler/profile_contraction_impl.hpp [hunk 2]
```diff
@@ -109,9 +109,9 @@ int profile_contraction_impl(ck::index_t do_verification,
     const std::vector<index_t> e_ms_ns_lengths = {M[0], M[1], N[0], N[1]};
     const std::vector<index_t> d_m_n_lengths   = {M[0], M[1], N[0], N[1]};
 
-    // The order of dims in StridesB is [K0, K1, N0, N1] so need to change it to [N0, N1, K0, K1]
-    const std::vector<index_t> b_ns_ks_strides = {
-        StridesB[2], StridesB[3], StridesB[0], StridesB[1]};
+    // // The order of dims in StridesB is [K0, K1, N0, N1] so need to change it to [N0, N1, K0, K1]
```
Comment (bartekxk) [inline profiler/include/profiler/profile_contraction_impl.hpp]: Remove one "//"

**Review by bartekxk — DISMISSED** [review]

**Review by bartekxk — APPROVED** [review]

**Review by LisaDelaney — COMMENTED** [review]
profiler/README.md looks good to me.

**Review by LisaDelaney — APPROVED** [review]

**Pull request merged**